feat(vertical nav): adds simple vertical nav list

mattnolting opened the pull request:
Closes #27

patternfly-build commented:
Deploy preview for *pf-next* ready!

Built with commit 63baeb4ebeee4923469ce1b834d394b346b72d0f

https://deploy-preview-311--pf-next.netlify.com

andresgalante commented:
Hi Matt, I still need to get your branch and review it, but to kick it off we need to remove the demo. is that just to preview it as we review?

mattnolting commented:
@andresgalante Removed demo and updated border decoration!

andresgalante commented on src/patternfly/components/VerticalNav/docs/code.md:
```diff
@@ -0,0 +1,24 @@
+## Overview
+
+The vertical navigation list contains list items, links, icons and text. Icons, along with text, must be included for the list items to align properly. Do not omit these elements. The chevron is required for list items containing subnavigation. 
```
@mcarrano can you confirm that icons are mandatory?

mcarrano replied:
Actually I don't think the icons should be mandatory, @mattnolting @andresgalante

andresgalante commented on src/patternfly/components/VerticalNav/docs/code.md:
```diff
@@ -0,0 +1,24 @@
+
+## Accessibility
+
+| Attribute | Applied To | Outcome |
```
Do we need role navigation to the `nav` element? is it necessary in modern browsers?

mattnolting replied:
Its use is suggested according to https://a11yproject.com/posts/aria-landmark-roles/

andresgalante commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,134 @@
+  /* stylelint-disable */
+
+  // Vertical nav
+  --pf-c-vertical-nav--Color:             var(--pf-global--BackgroundColor--dark-100);
```
Please let's not use a BackgroundColor to define a color. If we need more color options on a global level let me know

andresgalante commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,134 @@
+  background-color: var(--pf-c-vertical-nav--BackgroundColor);
+
+  // List item
+  &__item {
```
Maybe we can remove this, right?

mattnolting replied:
Yep

andresgalante commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,134 @@
+  &__link {
+    position: relative;
+    display: flex;
+    align-items: baseline;
```
Shouldn't this be `center`?

mattnolting replied:
I don't think so, it looks very odd when aligned center when text wraps

andresgalante commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,134 @@
+    padding-right: var(--pf-c-vertical-nav__link--PaddingRight);
+    padding-bottom: var(--pf-c-vertical-nav__link--PaddingBottom);
+    padding-left: var(--pf-c-vertical-nav__link--PaddingLeft);
+    margin-top: var(--pf-c-vertical-nav__link--MarginTop);
```
Do we need to change the margin? 
the base `[class^=pf-]` will make it `0` by default

mattnolting replied:
No, we don't. This is an assumption based on https://github.com/patternfly/patternfly-next/blob/dbf83c1f88b1097ff560c9e264a9aca7a0e411cb/patternfly_old/pf-c-vertical-nav/pf-c-vertical-nav.scss#L40-L41.

mattnolting replied:
removing margin vals

andresgalante commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,134 @@
+    margin-top: var(--pf-c-vertical-nav__link--MarginTop);
+    margin-bottom: var(--pf-c-vertical-nav__link--MarginBottom);
+    color: var(--pf-c-vertical-nav__link--Color);
+    text-decoration: none;
```
👍 to force `text-decoration: none;` here

andresgalante commented on src/patternfly/components/VerticalNav/docs/code.md:
```diff
@@ -0,0 +1,24 @@
+| `.pf-c-vertical-nav__link_icon` | `<span>` |  Initiates link icon. It is required. **Required** |
+| `.pf-c-vertical-nav__link_text` | `<span>` |  Initiates link text. It is required. **Required** |
+| `.pf-c-vertical-nav__link__arrow` | `<span>` |  Initiates link arrow, indicating subnavigation. |
+| `.pf-m-active` | `.pf-c-vertical-nav__link` |  Modifies for pf-c-vertical-nav__link :after, background-color and color. |
```
We are missing docs for disabled, hover and focus modifiers.

andresgalante commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,134 @@
+  
+  // Vertical nav link active :after
+  --pf-c-vertical-nav__link--after--active--Left:    calc(var(--pf-c-vertical-nav__link--PaddingLeft) + 1.5rem + var(--pf-c-vertical-nav__link__icon--MarginRight));
+  --pf-c-vertical-nav__link--after--active--Width:   2rem;
```
Can you please add a comment abut this magic values? I understand why they are here, but I wonder if people would.

mattnolting replied:
Removing the magic values

andresgalante replied:
not need to remove them, just make a comment to explain what they are doing

mattnolting replied:
sorry, replying on wrong comment

andresgalante commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,134 @@
+    // Icon
+    &__icon {
+      margin-right: var(--pf-c-vertical-nav__link__icon--MarginRight);
+      line-height: 1;
```
the need to find a way to make this global. Can you please add a comment? I'll create a story to fix the size of the icons somehow.

mattnolting replied:
Yep

andresgalante commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    // Arrow/chevron
+    &__arrow {
```
I would't absolute position the arrow, with flex we can center it to the other items without having to measure it like this. What do you think?

mattnolting replied:
Totally reasonable approach, my assumption here is that with or without the chevron, the link text width limit would align vertically, that that may not be necessary. Does that make sense? I would prefer to rely on flex.

jgiardino commented on src/patternfly/components/VerticalNav/vertical-nav-link-icon.hbs:
```diff
@@ -0,0 +1,3 @@
+<span class="pf-c-vertical-nav__link__icon {{ pf-c-vertical-nav--modifier }}">
```
In this case, since icons are required and potentially not coming from fontawesome, we should add `aria-hidden="true"` to the `<span>` so that this element is ignored by screen readers.

jgiardino commented on src/patternfly/components/VerticalNav/vertical-nav-link.hbs:
```diff
@@ -0,0 +1,10 @@
+<a href="{{url}}" class="pf-c-vertical-nav__link {{#if current}}pf-m-active {{/if}}{{ pf-c-vertical-nav--modifier }}" role="link" {{#if current}} aria-current="true" {{/if}}>
```
If this menu item has a submenu, then the `<a>` element should include `aria-expanded` set to `true` when expanded and `false` when collapsed.

jgiardino commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,134 @@
+    &:active,
+    &.pf-m-active {
+      color: var(--pf-c-vertical-nav__link--active--Color);
+      background-color: var(--pf-c-vertical-nav__link--active--BackgroundColor);
```
This color combination doesn't pass the color contrast check. @kybaker fyi...

This is where those colors are being applied:
![image](https://user-images.githubusercontent.com/21063328/41248515-48aa3e70-6d7f-11e8-98e2-3fbbf2ea8f10.png)

jgiardino commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,134 @@
+    &.pf-m-focus {
+      z-index: 2;
+      color: var(--pf-c-vertical-nav__link--focus--Color);
+      background-color: var(--pf-c-vertical-nav__link--focus--BackgroundColor);
```
Should `:active` and `:focus` be reordered? Currently, if I place focus on the Home link or click on it, it turns from blue to dark gray (er, whatever that color is). Is that the expected behavior, or did we want it to stay blue (@kybaker)?

mcarrano replied:
I took a look and I agree with @jgiardino . I would expect the active state to take precedence so that the text "Home" would remain blue.

mattnolting commented:
@andresgalante @jgiardino I've addressed all comments in this review.
Some notable changes:
- Padding has been moved from the button to the button text. This allows devs to omit a presentational icon and have the active border positioned correctly.
- Added a second conceptual layout for including a link AND button that opens a submenu. 
- Added aria labels throughout

andresgalante commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,140 @@
+  --pf-c-vertical-nav__link--hover--BackgroundColor:    var(--pf-global--BackgroundColor--light-200);
+  --pf-c-vertical-nav__link--focus--BackgroundColor:    var(--pf-global--BackgroundColor--light-200);
+  --pf-c-vertical-nav__link--active--BackgroundColor:   var(--pf-global--BackgroundColor--light-200);
+  --pf-c-vertical-nav__link--disabled--BackgroundColor: var(--pf-global--BackgroundColor--disabled);
```
This is an exception, from what I saw on the mockups it's white, lets use BackgroundColor--light-100

andresgalante replied:
This is an exception, from what I saw on the mockups it's white, lets use BackgroundColor--light-100

andresgalante commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    // Icon
+    &__icon {
```
I've just had a chat with @kybaker the new designs have no icons, please remove them, sorry

andresgalante commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,127 @@
+
+  // List
+  color: var(--pf-c-vertical-nav--Color);
+  background-color: var(--pf-c-vertical-nav--BackgroundColor);
```
Let's not assign a background to this component, since it'll take the white or light grey from the sidebar

mattnolting replied:
👍

andresgalante commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,127 @@
+    // Modifiers and states
+    &.pf-m-hover,
+    &:hover {
+      background-color: var(--pf-c-vertical-nav__link--hover--BackgroundColor);
```
I think the hover state should be the same as the active with the line decoration. @kybaker do you have a mockup for that?

andresgalante commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,127 @@
+    &.pf-m-focus,
+    &:focus {
+      z-index: 2;
+      background-color: var(--pf-c-vertical-nav__link--focus--BackgroundColor);
```
do we need to set the color here? is this to allow users to theme the component?

mattnolting replied:
Just a theming hook

andresgalante replied:
ok, even if it's a themeing hook I don't see the need for a `pf-m-focus` class

andresgalante commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,127 @@
+      background-color: var(--pf-c-vertical-nav__link--focus--BackgroundColor);
+
+      > * {
+        color: var(--pf-c-vertical-nav__link--focus--Color);
```
or here?

mattnolting replied:
Just a theming hook

andresgalante commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    &.pf-m-active,
+    &:active {
```
same here, do we actually need this?

mattnolting replied:
Just a theming hook

andresgalante commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,127 @@
+
+    &.pf-m-disabled,
+    &:disabled {
+      background-color: var(--pf-c-vertical-nav__link--disabled--BackgroundColor);
```
I don't think this is the color we should use, @kybaker do you have a mockup for a disabled item on the vertical menu?

mattnolting replied:
Should we even have a disabled style?

andresgalante commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    // Using this method to align active marker to text rather than link. Accomodates <a>'s without icons
+    &__text {
```
should this be `pf-c-vertical-nav__text` or `pf-c-vertical-nav__link-text` instead of `pf-c-vertical-nav__link__text` ?

mattnolting replied:
`pf-c-vertical-nav__link-text` Yes, it should

mattnolting commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,127 @@
+@import "../../patternfly-utilities";
+
+.pf-c-vertical-nav {
+  /* stylelint-disable */
```
We could, but allowing variable values to align vertically makes scannability sooo much better.

mattnolting replied:
Could we update stylelint config to allow this??

andresgalante replied:
I wouldn't change the linter since it'll allow a double space after the `:`. let me think about it.

mattnolting commented on src/patternfly/components/VerticalNav/styles.scss:
```diff
@@ -0,0 +1,127 @@
+  --pf-c-vertical-nav__link--hover--BackgroundColor:    var(--pf-global--BackgroundColor--light-100);
+  --pf-c-vertical-nav__link--focus--BackgroundColor:    var(--pf-global--BackgroundColor--light-100);
+  --pf-c-vertical-nav__link--active--BackgroundColor:   var(--pf-global--BackgroundColor--light-100);
+  --pf-c-vertical-nav__link--disabled--BackgroundColor: var(--pf-global--BackgroundColor--disabled);
```
👍
